File: workflow/scripts/step_6_combine_annotations/combine_constraint_anno.R
#!/usr/bin/env Rscript

# shell
# '''
# Rscript {input.script} \
#         -c {input.raw_snps} \
#         -n {input.filtered_snps} \
#         -f {input.derived_vars}
#         -g indexfile.txt \
#         -h {input.ancestral_fa} \
#         -i {input.parameter_log}
# '''

# module load R
# module load R_packages 
# gerp=results/annotation/gerp
# phylo=results/annotation/phast/phyloP
# phast=results/annotation/phast/phastCons
# index=results/annotation/indexfiles
# for i in {1..18} X; do Rscript combine_constraint_anno.R -c $i -n 30 -f $gerp -g $phast -h $phylo -i $index; done

# load packages
library(optparse)
library(stringr)
library(dplyr)
library(data.table)
library(tidyverse)

option_list = list(
  make_option(c("-c", "--chromosome"), type="character", default="1",
              help="chromosome number", metavar="character"),
  
  make_option(c("-n", "--n-chunks"), type = "integer", default = 30,
              help="number of chunks", metavar="integer "),
  
  # to do, make more generic and if else
  make_option(c("-f", "--phast-folder"), type="character", default="/Users/juliahoglund/Documents/localCADD/testdata/phastCons/",
              help="path to phastCons scores", metavar="character"),
  
  make_option(c("-g", "--phylo-folder"), type="character", default="/Users/juliahoglund/Documents/localCADD/testdata/phyloP/",
              help="path to phyloP scores", metavar="character"),
  
  make_option(c("-i", "--gerp-folder"), type="character", default="/Users/juliahoglund/Documents/localCADD/testdata/gerp/",
              help="path to gerp scores", metavar="character"),
  
  make_option(c("-j", "--index-folder"), type="character", default="/Users/juliahoglund/Documents/localCADD/testdata/indexfiles/",
              help="path to bp position index files", metavar="character")  
)

opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

for (i in 1:opt$`n-chunks`) {
  
  message('opening constraint annotations..', opt$chromosome, ':', i, '/', opt$`n-chunks`)
  
  phastC <- fread(paste(opt$`phast-folder`, 'chr', opt$chromosome, '/chr', opt$chromosome, '_', i, '.phast.bed', sep=""), drop=c(1,4), header=FALSE) %>% 
    rename('start' = V2, 'end' = V3, 'phastCons' = V5)
  
  phyloP <- fread(paste(opt$`phylo-folder`, 'chr', opt$chromosome, '/chr', opt$chromosome, '_', i, '.phylo.bed', sep = ""), drop=c(1,4), header=FALSE) %>% 
    rename('start' = V2, 'end' = V3, 'phyloP' = V5)
  
  gerp  <- fread(paste(opt$`gerp-folder`, 'chr', opt$chromosome, '/chr', opt$chromosome, '_', i, '.rates.parsed', sep = ""), header=FALSE) %>% 
    rename('GERP_NS' = V1, 'GERP_RS' = V2)
  
  index <- 
    fread(paste(opt$`index-folder`, 'chr', opt$chromosome, '/chr', opt$chromosome, '_', i, '.index', sep = ""), header = FALSE) %>% 
    separate_wider_delim(cols=V1, delim=": ", names=c('V1', 'start')) %>% 
    separate_wider_delim(cols=V2, delim=": ", names=c('V2', 'size')) %>% 
    dplyr::select(start, size) %>% 
    mutate(start = as.numeric(start), size = as.numeric(size), end = start+size-1) %>% 
    relocate(end, .before = size)
  
  message("creating vector of base pair positions..")
  
  start <- vector()
  
  for (j in 1:nrow(index)) {
    start <- append(start, (seq(index$start[j], index$end[j], 1)))
  }
  
  gerp <- cbind(start, gerp)
  
  phast <- full_join(phastC, phyloP) %>% 
    mutate(chr = paste("chr", opt$chromosome, sep = "")) %>% 
    relocate(chr, .before = start) %>%
    mutate(end = start)
  ## with or without CHR in bed???
  
  
  message("combining and writing file .. ", i) 
  constraint <- full_join(gerp, phast)
  
  write.table(constraint, file=paste("constraint.", opt$chromosome, "_", i, ".bed", sep = ""), col.names = TRUE, row.names = FALSE, quote = FALSE)
  
}

File: workflow/scripts/step_6_combine_annotations/combine_constraint_anno.py
import argparse
import glob
import os
import pandas as pd
import sys

def parse_args():
    parser = argparse.ArgumentParser(description="Combine constraint annotations for a chromosome.")
    parser.add_argument("-c", "--chromosome", type=str, required=True, help="chromosome number")
    parser.add_argument("-f", "--phast_folder", type=str, default="results/annotation/phast/phastCons", help="path to phastCons scores")
    parser.add_argument("-g", "--phylo_folder", type=str, default="results/annotation/phast/phyloP", help="path to phyloP scores")
    parser.add_argument("-i", "--gerp_folder", type=str, default="results/annotation/gerp", help="path to gerp scores")
    parser.add_argument("-o", "--output", type=str, default=None, help="output file name (default: constraint.<chromosome>.bed)")
    return parser.parse_args()

def load_and_concat(folder, chrom, suffix):
    # Try both: files directly in folder and in folder/chrN/
    files = []
    # 1. Check in folder/chrN/
    chr_folder = os.path.join(folder, f"chr{chrom}")
    if os.path.isdir(chr_folder):
        all_chr_files = os.listdir(chr_folder)
        if suffix == "phast.bed":
            files += [os.path.join(chr_folder, f) for f in all_chr_files if f.startswith(f"chr{chrom}_multiway-") and f.endswith(".phast.bed")]
        elif suffix == "phylo.bed":
            files += [os.path.join(chr_folder, f) for f in all_chr_files if f.startswith(f"chr{chrom}_multiway-") and f.endswith(".phylo.bed")]
    # 2. Also check in the main folder
    try:
        all_files = os.listdir(folder)
        if suffix == "phast.bed":
            files += [os.path.join(folder, f) for f in all_files if f.startswith(f"chr{chrom}_multiway-") and f.endswith(".phast.bed")]
        elif suffix == "phylo.bed":
            files += [os.path.join(folder, f) for f in all_files if f.startswith(f"chr{chrom}_multiway-") and f.endswith(".phylo.bed")]
    except FileNotFoundError:
        print(f"Directory {folder} does not exist.")
    # 3. Fallback for other suffixes (e.g. GERP)
    if not files and suffix not in ("phast.bed", "phylo.bed"):
        pattern = os.path.join(folder, f"chr{chrom}", f"chr{chrom}-*.{suffix}")
        print(f"Looking for files with pattern: {pattern}")
        files = sorted(glob.glob(pattern))
    # Sort files by chunk number (assumes format chrN_multiway-<chunk>.<suffix>)
    def extract_chunk_number(filename):
        import re
        match = re.search(r'_multiway-(\d+)\.', filename)
        return int(match.group(1)) if match else -1
    files = sorted(files, key=extract_chunk_number)
    print(f"Filtered and sorted files for {suffix}")
    if not files:
        print(f"No files found for {suffix} in {folder} or {chr_folder}")
    dfs = []
    for f in files:
        print(f"Reading file: {f}")
        df = pd.read_csv(f, sep="\t", header=None)
        dfs.append(df)
    return pd.concat(dfs, ignore_index=True) if dfs else pd.DataFrame()

def main():
    args = parse_args()
    chrom = args.chromosome

    print("Loading and combining phastCons files...")
    phastC = load_and_concat(args.phast_folder, chrom, "phast.bed")
    if phastC.empty:
        raise RuntimeError("No phastCons files found.")
    if phastC.shape[1] == 5:
        phastC = phastC.drop([0,3], axis=1)
        phastC.columns = ['start', 'end', 'phastCons']
    elif phastC.shape[1] == 3:
        phastC.columns = ['start', 'end', 'phastCons']
    elif phastC.shape[1] == 4:
        phastC = phastC.iloc[:, [1,2,3]]
        phastC.columns = ['start', 'end', 'phastCons']
    else:
        raise ValueError(f"Unexpected number of columns in phastCons input: {phastC.shape[1]}")

    print("Loading and combining phyloP files...")
    phyloP = load_and_concat(args.phylo_folder, chrom, "phylo.bed")
    if phyloP.empty:
        raise RuntimeError("No phyloP files found.")
    if phyloP.shape[1] == 5:
        phyloP = phyloP.drop([0,3], axis=1)
        phyloP.columns = ['start', 'end', 'phyloP']
    elif phyloP.shape[1] == 3:
        phyloP.columns = ['start', 'end', 'phyloP']
    elif phyloP.shape[1] == 4:
        phyloP = phyloP.iloc[:, [1,2,3]]
        phyloP.columns = ['start', 'end', 'phyloP']
    else:
        raise ValueError(f"Unexpected number of columns in phyloP input: {phyloP.shape[1]}")

    print("Merging phastCons and phyloP...")
    phast = pd.merge(phastC, phyloP, on=['start', 'end'], how='outer')
    phast['chr'] = f"chr{chrom}"
    phast = phast[['chr', 'start', 'end', 'phastCons', 'phyloP']]
    phast = phast.reset_index(drop=True)
    phast['Pos'] = phast.index + 1

    print("Finding and loading GERP file...")
    gerp_file = None
    for f in os.listdir(args.gerp_folder):
        if f"chr{chrom}" in f and f.endswith(".parsed"):
            gerp_file = os.path.join(args.gerp_folder, f)
            break
    if not gerp_file:
        raise RuntimeError(f"No GERP .parsed file found for chromosome {chrom} in {args.gerp_folder}")

    gerp = pd.read_csv(gerp_file, sep="\t", header=None, names=['Pos', 'GERP_NS', 'GERP_RS'])

    print("Merging all annotations on Pos...")
    constraint = pd.merge(phast, gerp, on='Pos', how='inner')
    # Drop start and end columns after merge
    constraint = constraint.drop(columns=['start', 'end'])
    # Rename chromosome column to #Chrom, remove 'chr' prefix, and make integer
    constraint = constraint.rename(columns={'chr': '#Chrom'})
    constraint['#Chrom'] = constraint['#Chrom'].astype(str).str.replace('chr', '', regex=False).astype(int)
    constraint = constraint[['#Chrom', 'Pos', 'GERP_NS', 'GERP_RS', 'phastCons', 'phyloP']]

    print("Writing output...")
    output_file = args.output or f"constraint.{chrom}.bed"
    constraint.to_csv(output_file, sep="\t", index=False, header=True, float_format="%.6f")
    print("Done.")

if __name__ == "__main__":
    main()
